mos-platform/common/include/fixed_point.h

```diff
@@ -0,0 +1,301 @@
+    val -= o.val;
+    return *this;
+  }
+
+  template<typename T, typename = typename std::enable_if<std::is_arithmetic<T>::value, T>::type>
+  [[clang::always_inline]] constexpr Value operator/(T i) {
+    Value<ISize, FSize> n{val};
+    n /= i;
+    return n;
+  }
+  template<typename T, typename = typename std::enable_if<std::is_arithmetic<T>::value, T>::type>
+  [[clang::always_inline]] constexpr Value& operator/=(T i) {
+    val /= i;
+    return *this;
+  }
+
+  template<typename T, typename = typename std::enable_if<std::is_arithmetic<T>::value, T>::type>
+  [[clang::always_inline]] constexpr Value operator*(T i) {
+    Value<ISize, FSize> n{val};
+    n *= i;
+    return n;
+  }
+  template<typename T, typename = typename std::enable_if<std::is_arithmetic<T>::value, T>::type>
+  [[clang::always_inline]] constexpr Value& operator*=(T i) {
+    val *= i;
+    return *this;
+  }
+
+  template<typename T, typename = typename std::enable_if<std::is_arithmetic<T>::value, T>::type>
+  [[clang::always_inline]] constexpr Value operator >>(T v) const {
+    Value<ISize, FSize> n = *this;
+    n >>= v;
+    return n;
+  }
+  template<typename T, typename = typename std::enable_if<std::is_arithmetic<T>::value, T>::type>
+  [[clang::always_inline]] constexpr Value& operator >>=(T v) {
+    val >>= v;
+    return *this;
+  }
+  
+  template<typename T, typename = typename std::enable_if<std::is_arithmetic<T>::value, T>::type>
+  [[clang::always_inline]] constexpr Value operator <<(T v) const {
+    Value<ISize, FSize> n = *this;
+    n <<= v;
+    return n;
+  }
+  template<typename T, typename = typename std::enable_if<std::is_arithmetic<T>::value, T>::type>
+  [[clang::always_inline]] constexpr Value& operator <<=(T v) {
+    val <<= v;
+    return *this;
+  }
+
+  [[clang::always_inline]] Value &operator++() {
+    val += 1;
+    return *this;
+  }
+  [[clang::always_inline]] Value operator++(int) {
+    Value old = *this;
+    ++*this;
+    return old;
+  }
+
+  // Comparison overloads
+
+  [[clang::always_inline]] constexpr friend bool operator==(const Value& a, const Value& b) noexcept {
+    return a.val == b.val;
+  }
+  [[clang::always_inline]] constexpr friend bool operator< (const Value& a, const Value& b) noexcept {
+    return a.val < b.val;
+  }
+  [[clang::always_inline]] constexpr friend bool operator!=(const Value& a, const Value& b) noexcept {
+    return !(a == b);
+  }
+  [[clang::always_inline]] constexpr friend bool operator> (const Value& a, const Value& b) noexcept {
+    return b < a;
+  }
+  [[clang::always_inline]] constexpr friend bool operator>=(const Value& a, const Value& b) noexcept {
+    return !(a < b);
+  }
+  [[clang::always_inline]] constexpr friend bool operator<=(const Value& a, const Value& b) noexcept {
+    return !(b < a);
+  }
+
+  [[clang::always_inline]] constexpr friend bool operator==(const Value& a, IntType b) noexcept {
+    return a.val == b;
```

Comment:
These aren't correct; they don't mask away the fractional bits or shift the integer to put it in line with the integer bits.